Inline Latex Previews

zachberry opened the pull request:
Requires #1159 
Fixes #1206

keeeeeegan commented:
I noticed something. If your selection of text includes other formatting (such as sup text) and then click the function button, a blank page loads and you can't go back. I had this happen also for headings.

![2020-03-31 15 02 38](https://user-images.githubusercontent.com/229935/78065209-edafd680-7360-11ea-8d66-88e86b7e515b.gif)

I also tried this with 9 and didn't see this issue.

iturgeon commented:
Something strange is going on w/ this PR - too many file changes (they include commits already in dev10)

qwertynerd97 commented:
The weird repeated file changes have been fixed by merging emerald into the PR!

keeeeeegan commented:
Checked again, looks good!

zachberry commented:
Made a few changes:

* The preview wasn't showing up when (for example) put in Question explanations. Increased the z-index. Also moved where HoveringPreview exists in the JSX such that it will be under the toolbar at the top
* Previews weren't moving when scrolling the page - moving where the component is in the JSX also fixed this. Had to tweak the way it positioned itself, requiring a reference to the editor-page DOM element to get working
* When resizing the page the preview wasn't moving. Added a ResizeObserver to listen for this. ResizeObserver doesn't work in Safari, but the alternative is not worth it imho, and this is somewhat obscure and won't break anything if the preview doesn't move when resizing.

iturgeon commented on packages/app/obojobo-document-engine/__tests__/oboeditor/components/page-editor.test.js:
```diff
@@ -25,8 +25,9 @@ jest.mock('src/scripts/oboeditor/stores/editor-store', () => ({
 jest.mock('src/scripts/oboeditor/components/toolbars/paragraph-styles')
-// jest.mock('src/scripts/oboeditor/components/toolbars/content-toolbar')
-//jest.mock('obojobo-document-engine/src/scripts/common/registry')
+jest.mock('src/scripts/oboeditor/components/hovering-preview', () => props => (
```
I added a mock for hovering-preview here to avoid rendering it in this test

iturgeon commented on packages/app/obojobo-document-engine/__tests__/oboeditor/components/hovering-preview.test.js:
```diff
@@ -0,0 +1,176 @@
+jest.mock('slate-react')
+jest.mock('slate')
+
+describe('HoveringPreview', () => {
```
Added a test just for HoveringPreview

iturgeon commented on packages/app/obojobo-document-engine/src/scripts/oboeditor/components/hovering-preview.js:
```diff
@@ -0,0 +1,70 @@
+import './hovering-preview.scss'
+const LATEX_MARK = '_latex'
+
+const HoveringPreview = ({ pageEditorContainerRef }) => {
```
Named props

iturgeon commented on packages/app/obojobo-document-engine/src/scripts/oboeditor/components/hovering-preview.js:
```diff
@@ -0,0 +1,70 @@
+
+	// SHOW, HIDE, AND UPDATE LOCATION ON SCREEN
+	// only runs when text changes OR window width changes
+	useLayoutEffect(() => {
```
Switched to `useLayoutEffect` since this changes the dom element's style

iturgeon commented on packages/app/obojobo-document-engine/src/scripts/oboeditor/components/hovering-preview.js:
```diff
@@ -0,0 +1,70 @@
+		el.style.opacity = 1
+		el.style.top = `${rect.top - pageEditorRect.top - el.offsetHeight - 6}px`
+		el.style.left = `${rect.left - pageEditorRect.left - el.offsetWidth / 2 + rect.width / 2}px`
+	}, [window.innerWidth, leaf.text, leaf])
```
Added conditional watchers to useLayoutEffect - this should now only render when one of these changes

* window.innerWidth - no need to re-render on height changes as they have no effect on our calcs
* leaf.text - any time the leaf node text changes - this can be more often then desired because whatever is selected will feed it's own text in, latex or not
* leaf - if the leaf object itself is different (needed to recalc the location for 2 different equations w/ the same text)

iturgeon commented on packages/app/obojobo-document-engine/src/scripts/oboeditor/components/hovering-preview.js:
```diff
@@ -0,0 +1,70 @@
+
+	// RENDER KATEX HTML
+	// run only when text changes
+	const katexHTML = useMemo(() => {
+		if (!leaf[LATEX_MARK]) return ''
+		return katex.renderToString(leaf.text, { throwOnError: false })
+	}, [leaf.text])
```
katex rendering is fairly costly - so I added useMemo here only calculate when leaf text changes.

Another limitation - it'll run as the selection changes, even if its not a katex leaf.  So I added an additional quick exit inside the function if the mark isn't a latex mark

zachberry commented:
@iturgeon Updates change the behavior when you use the arrow keys to move into a inline latex equation. Before the preview would be positioned as expected:

![2020-04-14 13 29 47](https://user-images.githubusercontent.com/73479/79255399-3af97100-7e54-11ea-88b2-732be66ac7cb.gif)

With the updates clicking works but using the keyboard to move the cursor inside the equation results in the preview window showing up at the wrong location:

![2020-04-14 13 30 17](https://user-images.githubusercontent.com/73479/79255425-46e53300-7e54-11ea-9360-b760f9f120a3.gif)

zachberry commented:
In hovering-preview switching `useLayoutEffect` for `useEffect` does fix the positioning problem, if that doesn't bog down our rendering too much

iturgeon commented:
Oh interesting, I obviously wasn't using arrow keys.  I'll have to think about the differences between useEffect and useLayoutEffect more

iturgeon commented:
Conflicts solved and switched back to useEffect after some reading.

iturgeon commented:
Hovering preview is above the more details menus.

<img width="512" alt="Screen Shot 2020-04-16 at 10 23 30 AM" src="https://user-images.githubusercontent.com/73480/79467977-6737fd80-7fcc-11ea-8931-663a3f3a0f69.png">

zachberry commented:
I can get the More Info Box to show up over the inline preview in basic nodes like texts and lists, but inside a Question (for example) the stacking contexts are different and the preview still shows up over the box.

I can hack around this by changing the focus into the MoreInfoBox:

![2020-04-17 11 05 59](https://user-images.githubusercontent.com/73479/79583882-a2553200-809b-11ea-9933-3c0b83169564.gif)

It's also nice since keyboard focus is now inside the box. However, does this pose an issue with accessibility since the slate selection is lost? I was thinking we could capture the slate selection and restore it on close, but I'm not sure how (and if that is possible). Summoning @qwertynerd97

qwertynerd97 commented:
Yeah, I actually just finished setting the more info box to focus on open and restore selection on close on my keyboard accessibility branch.  I've made a draft PR so you can take a look over at #1268

zachberry commented:
In this case, I'm thinking we could let the layer issue ride for now and fix when #1268 gets merged in
